src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -237,84 +243,112 @@ public IEnumerable<char> PotentialCommitCharacters
                 return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.None);
             }
 
-            var textChange = change.TextChange;
-            var triggerSnapshotSpan = new SnapshotSpan(triggerSnapshot, textChange.Span.ToSpan());
-            var mappedSpan = triggerSnapshotSpan.TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
-
-            using (var edit = subjectBuffer.CreateEdit(EditOptions.DefaultMinimalChange, reiteratedVersionNumber: null, editTag: null))
+            // Specifically for snippets, we use reflection to try and invoke the LanguageServerSnippetExpander's
+            // TryExpand method and determine if it succeeded or not.
+            var snippetServiceSucceeded = false;
+            if (SnippetCompletionItem.IsSnippet(roslynItem))
             {
-                edit.Replace(mappedSpan.Span, change.TextChange.NewText);
-
-                // edit.Apply() may trigger changes made by extensions.
-                // updatedCurrentSnapshot will contain changes made by Roslyn but not by other extensions.
-                var updatedCurrentSnapshot = edit.Apply();
+                var type = _languageServerSnippetExpander.GetType();
+                var expandMethod = type.GetMethod("TryExpand");
+                var lspSnippetText = change.LSPSnippet;
+                var sourceText = document.GetTextAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+                var textEdit = new LSP.TextEdit()
+                {
+                    Range = ProtocolConversions.TextSpanToRange(change.TextChange.Span, sourceText),
+                    NewText = lspSnippetText!
+                };
 
-                if (change.NewPosition.HasValue)
+                if (expandMethod is not null)
                 {
-                    // Roslyn knows how to position the caret in the snapshot we just created.
-                    // If there were more edits made by extensions, TryMoveCaretToAndEnsureVisible maps the snapshot point to the most recent one.
-                    view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(updatedCurrentSnapshot, change.NewPosition.Value));
+                    var expandMethodResult = expandMethod.Invoke(_languageServerSnippetExpander, new object[] { textEdit, view, triggerSnapshot });
+                    snippetServiceSucceeded = expandMethodResult is not null && (bool)expandMethodResult;
                 }
-                else
+            }
+
+            // For all other completions and if the snippet LSP service fails
+            // we still want to try and insert the snippet sans the placeholders + tab stop behavior.
+            if (!snippetServiceSucceeded)
```

Comment:
i don't get this boolean.  if we have a snippet, it doesn't matter if it succeeds or not, we shoudln't fallback to doing something random.